File: app/src/main/java/com/theandroiddev/riffking/HomeFragment.java
package com.theandroiddev.riffking;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;


/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * {@link HomeFragment.OnFragmentInteractionListener} interface
 * to handle interaction events.
 * Use the {@link HomeFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class HomeFragment extends Fragment {
    public static final String KEY_LAYOUT_MANAGER = "mLayoutManager";
    private static final String TAG = "HomeFragment";

    // TODO: Rename parameter arguments, choose names that match

    protected RecyclerView mRecyclerView;
    protected ThreadAdapter mThreadAdapter;
    protected RecyclerView.LayoutManager mLayoutManager;
    protected ArrayList<Thread> threads;
    protected ArrayList<Thread> threadsToRemove;
    protected LayoutManagerType mCurrentLayoutManagerType;
    protected Helper helper;
    ItemTouchHelper.SimpleCallback simpleCallback = new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT) {

        @Override
        public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
            return false;
        }

        @Override
        public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {

            int pos = viewHolder.getAdapterPosition();
            toRemove(pos);

        }
    };
    private FirebaseDatabase database = FirebaseDatabase.getInstance();
    private DatabaseReference mDatabase;
    private OnFragmentInteractionListener mListener;
    private LayoutManagerType recyclerViewLayoutManager;

    public HomeFragment() {
        // Required empty public constructor
    }

    private void toRemove(int position) {
        showUndoSnackbar(position);

    }

    private void showUndoSnackbar(final int position) {
        final Thread thread = threads.get(position);
        Snackbar snackbar = Snackbar
                .make(mRecyclerView, "REMOVED", Snackbar.LENGTH_LONG)
                .setAction("UNDO", new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        undoToRemove(thread, position);
                    }
                });
        snackbar.show();

        sendToRemoveQueue(thread, position);
    }

    private void sendToRemoveQueue(Thread thread, int position) {
        threads.remove(position);
        mThreadAdapter.notifyItemRemoved(position);
        threadsToRemove.add(thread);
    }

    private void undoToRemove(Thread thread, int position) {
        threads.add(position, thread);
        mThreadAdapter.notifyItemInserted(position);
        mRecyclerView.scrollToPosition(position);
        threadsToRemove.remove(thread);
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param param1 Parameter 1.
     * @param param2 Parameter 2.
     * @return A new instance of fragment HomeFragment.
     */
    // TODO: Rename and change types and number of parameters
    public HomeFragment newInstance(String param1, String param2, String param3, String param4) {
        HomeFragment fragment = new HomeFragment();
        Bundle args = new Bundle();

        fragment.setArguments(args);

        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mDatabase = FirebaseDatabase.getInstance().getReference();
        helper = new Helper();

        threads = new ArrayList<>();
        threadsToRemove = new ArrayList<>();

        FirebaseDatabase database = FirebaseDatabase.getInstance();
        DatabaseReference databaseReference = database.getReference();
        databaseReference.child("threads").addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                threads.clear();
                Iterable<DataSnapshot> children = dataSnapshot.getChildren();

                for (DataSnapshot child : children) {
                    Thread thread = child.getValue(Thread.class);
                    thread.setId(child.getKey());
                    threads.add(0, thread);

                }
                mThreadAdapter.notifyDataSetChanged();
                Log.e(TAG, "onDataChange:QWE ");

            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });

    }

    @Override
    public void onPause() {
        super.onPause();
        remove();

    }

    private void remove() {


        while (threadsToRemove.size() > 0) {
            Log.d(TAG, "remove: " + threadsToRemove + " " + threadsToRemove.size());

            mDatabase.child("threads").child(threadsToRemove.get(0).getId()).removeValue();
            threadsToRemove.remove(0);

        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View rootView = inflater.inflate(com.theandroiddev.riffking.R.layout.fragment_home, container, false);
        rootView.setTag(TAG);

        mRecyclerView = (RecyclerView) rootView.findViewById(R.id.recyclerView);


        mLayoutManager = new LinearLayoutManager(getActivity());
        mCurrentLayoutManagerType = LayoutManagerType.LINEAR_LAYOUT_MANAGER;
        if (savedInstanceState != null) {
            // Restore saved layout manager type.
            mCurrentLayoutManagerType = (LayoutManagerType) savedInstanceState
                    .getSerializable(KEY_LAYOUT_MANAGER);
        }
        setRecyclerViewLayoutManager(mCurrentLayoutManagerType);
        mThreadAdapter = new ThreadAdapter(getContext(), threads, mDatabase);
        mRecyclerView.setAdapter(mThreadAdapter);

        if (HomeActivity.user.getEmail().equals("jakubpchmiel@gmail.com")) {
            ItemTouchHelper itemTouchHelper = new ItemTouchHelper(simpleCallback);
            itemTouchHelper.attachToRecyclerView(mRecyclerView);
        }

        //TODO check why set two times
        setRecyclerViewLayoutManager(LayoutManagerType.LINEAR_LAYOUT_MANAGER);


        return rootView;
    }

    public void setRecyclerViewLayoutManager(LayoutManagerType layoutManagerType) {
        int scrollPosition = 0;

        // If a layout manager has already been set, get current scroll position.
        if (mRecyclerView.getLayoutManager() != null) {
            scrollPosition = ((LinearLayoutManager) mRecyclerView.getLayoutManager())
                    .findFirstCompletelyVisibleItemPosition();
        }

        switch (layoutManagerType) {
            case GRID_LAYOUT_MANAGER:
                mLayoutManager = new GridLayoutManager(getActivity(), Helper.SPAN_COUNT);
                mCurrentLayoutManagerType = LayoutManagerType.GRID_LAYOUT_MANAGER;
                break;
            case LINEAR_LAYOUT_MANAGER:
                mLayoutManager = new LinearLayoutManager(getActivity());
                mCurrentLayoutManagerType = LayoutManagerType.LINEAR_LAYOUT_MANAGER;
                break;
            default:
                mLayoutManager = new LinearLayoutManager(getActivity());
                mCurrentLayoutManagerType = LayoutManagerType.LINEAR_LAYOUT_MANAGER;
        }

        mRecyclerView.setLayoutManager(mLayoutManager);
        mRecyclerView.scrollToPosition(scrollPosition);
    }

    @Override
    public void onSaveInstanceState(Bundle savedInstanceState) {
        // Save currently selected layout manager.
        savedInstanceState.putSerializable(KEY_LAYOUT_MANAGER, mCurrentLayoutManagerType);
        super.onSaveInstanceState(savedInstanceState);
    }

    private boolean userIsConnected() {

        return Utility.isNetworkAvailable(getContext());
//        String textt = "";
//
//        if(mFirebaseAuth.isC() != null) {
//            textt = mFirebaseAuth.getCurrentUser().getEmail();
//        }
//
//        Toast.makeText(getContext(),textt , Toast.LENGTH_SHORT).show();

    }
    @Override
    public void onResume() {
        super.onResume();

        if (!userIsConnected()) {
            Toast.makeText(getContext(), "No internet connection", Toast.LENGTH_SHORT).show();
        }


        mThreadAdapter.notifyDataSetChanged();

        getView().setFocusableInTouchMode(true);
        getView().requestFocus();
//        getView().setOnKeyListener(new View.OnKeyListener() {
//            @Override
//            public boolean onKey(View v, int keyCode, KeyEvent event) {
//
//                if (event.getAction() == KeyEvent.ACTION_UP && keyCode == KeyEvent.KEYCODE_BACK) {
//
//                    getActivity().finish();
//
//                    return true;
//                }
//
//
//                return false;
//            }
//        });

        //TODO shoudnt it be in onCreateView?

        HomeActivity homeActivity = (HomeActivity) getActivity();
        homeActivity.fab.setVisibility(View.VISIBLE);
        homeActivity.fab.setImageResource(R.drawable.ic_add_24dp);
        homeActivity.fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                insertThread();
            }
        });
    }

    private void insertThread() {

        InsertThreadFragment insertThreadFragment = new InsertThreadFragment();

        FragmentManager fragmentManager = ((FragmentActivity) getContext()).getSupportFragmentManager();
        FragmentTransaction transaction = fragmentManager.beginTransaction();
        transaction.replace(com.theandroiddev.riffking.R.id.content_home, insertThreadFragment).addToBackStack(null);
        transaction.commit();

    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);

        if (isVisibleToUser && isResumed()) {
            onResume();
        }
    }

    // TODO: Rename method, update argument and hook method into UI event
    public void onButtonPressed(Uri uri) {
        if (mListener != null) {
            mListener.onFragmentInteraction(uri);
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }


    public enum LayoutManagerType {
        GRID_LAYOUT_MANAGER,
        LINEAR_LAYOUT_MANAGER
    }


    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnFragmentInteractionListener {
        // TODO: Update argument type and name
        void onFragmentInteraction(Uri uri);
    }
}

File: app/src/main/java/com/theandroiddev/riffking/presentation/home/HomeFragment.kt
package com.theandroiddev.riffking.presentation.home

import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.google.firebase.database.*
import com.theandroiddev.riffking.Helper
import com.theandroiddev.riffking.R
import com.theandroiddev.riffking.Utility
import com.theandroiddev.riffking.presentation.thread.Thread
import com.theandroiddev.riffking.presentation.thread.ThreadAdapter
import com.theandroiddev.riffking.presentation.upload.UploadFragment
import kotlinx.android.synthetic.main.fragment_home.*
import java.util.*

class HomeFragment : Fragment() {

    private var threadAdapter: ThreadAdapter? = null
    private var layoutManager = LinearLayoutManager(activity)
    private var currentLayoutManagerType = LayoutManagerType.LINEAR_LAYOUT_MANAGER
    private var threads: ArrayList<Thread> = arrayListOf()
    private var threadsToRemove: ArrayList<Thread> = arrayListOf()
    var helper: Helper? = null

    init {
        val context = context
        if (context != null) {
            helper = Helper(context)
        }
    }

    private var simpleCallback: ItemTouchHelper.SimpleCallback = object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT) {

        override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
            return false
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {

            val pos = viewHolder.adapterPosition
            toRemove(pos)

        }
    }

    private var database: DatabaseReference? = null
    private var listener: OnFragmentInteractionListener? = null

    private fun toRemove(position: Int) {
        showUndoSnackbar(position)
    }

    private fun showUndoSnackbar(position: Int) {
        val thread = threads[position]
        val snackbar = Snackbar
                .make(recyclerView, "REMOVED", Snackbar.LENGTH_LONG)
                .setAction("UNDO") { undoToRemove(thread, position) }
        snackbar.show()

        sendToRemoveQueue(thread, position)
    }

    private fun sendToRemoveQueue(thread: Thread, position: Int) {
        threads.removeAt(position)
        threadAdapter?.notifyItemRemoved(position)
        threadsToRemove.add(thread)
    }

    private fun undoToRemove(thread: Thread, position: Int) {
        threads.add(position, thread)
        threadAdapter?.notifyItemInserted(position)
        recyclerView.scrollToPosition(position)
        threadsToRemove.remove(thread)
    }

    fun newInstance(param1: String, param2: String, param3: String, param4: String): HomeFragment {
        val fragment = HomeFragment()
        val args = Bundle()

        fragment.arguments = args

        return fragment
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        database = FirebaseDatabase.getInstance().reference

        val database = FirebaseDatabase.getInstance()
        val databaseReference = database.reference
        databaseReference.child("threads").addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                threads.clear()
                val children = dataSnapshot.children

                for (child in children) {
                    val thread = child.getValue(Thread::class.java)
                    thread?.id = child.key
                    if (thread != null) {
                        threads.add(0, thread)
                    }
                }
                threadAdapter?.notifyDataSetChanged()
                Log.e(TAG, "onDataChange:QWE ")

            }

            override fun onCancelled(databaseError: DatabaseError) {

            }
        })

    }

    override fun onPause() {
        super.onPause()
        remove()
    }

    private fun remove() {

        while (threadsToRemove.size > 0) {
            Log.d(TAG, "remove: " + threadsToRemove + " " + threadsToRemove.size)

            val threadToRemoveId = threadsToRemove[0].id
            if (threadToRemoveId != null) {
                database?.child("threads")?.child(threadToRemoveId)?.removeValue()
                threadsToRemove.removeAt(0)
            } else {
                //TODO error
            }

        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val rootView = inflater.inflate(R.layout.fragment_home, container, false)
        rootView.tag = TAG

        if (savedInstanceState != null) {
            // Restore saved layout manager type.
            currentLayoutManagerType = savedInstanceState
                    .getSerializable(KEY_LAYOUT_MANAGER) as LayoutManagerType
        }
        val context = context
        val database = database
        val activity = activity
        if (context != null && database != null && activity != null) {
            threadAdapter = ThreadAdapter(context, threads, database)
            recyclerView.adapter = threadAdapter

            recyclerView.setHasFixedSize(true)
            recyclerView.setItemViewCacheSize(20)
            recyclerView.isDrawingCacheEnabled = true
            recyclerView.drawingCacheQuality = View.DRAWING_CACHE_QUALITY_HIGH

            if (HomeActivity.user?.email.equals("jakubpchmiel@gmail.com")) {
                val itemTouchHelper = ItemTouchHelper(simpleCallback)
                itemTouchHelper.attachToRecyclerView(recyclerView)
            }

            helper?.setRecyclerViewLayoutManager(LayoutManagerType.LINEAR_LAYOUT_MANAGER,
                    recyclerView, activity, layoutManager, currentLayoutManagerType)

            threadAdapter?.notifyDataSetChanged()
        }
        return rootView
    }

    override fun onSaveInstanceState(savedInstanceState: Bundle) {
        // Save currently selected layout manager.
        savedInstanceState.putSerializable(KEY_LAYOUT_MANAGER, currentLayoutManagerType)
        super.onSaveInstanceState(savedInstanceState)
    }

    private fun userIsConnected(): Boolean {
        val context = context
        return if(context != null) {
            Utility.isNetworkAvailable(context)
        } else false
        //        String textt = "";
        //
        //        if(firebaseAuth.isC() != null) {
        //            textt = firebaseAuth.getCurrentUser().getEmail();
        //        }
        //
        //        Toast.makeText(getContext(),textt , Toast.LENGTH_SHORT).show();

    }

    override fun onResume() {
        super.onResume()

        if (!userIsConnected()) {
            Toast.makeText(context, "No internet connection", Toast.LENGTH_SHORT).show()
        }

        //threadAdapter?.notifyDataSetChanged();

        view?.isFocusableInTouchMode = true
        view?.requestFocus()
        //        getView().setOnKeyListener(new View.OnKeyListener() {
        //            @Override
        //            public boolean onKey(View v, int keyCode, KeyEvent event) {
        //
        //                if (event.getAction() == KeyEvent.ACTION_UP && keyCode == KeyEvent.KEYCODE_BACK) {
        //
        //                    getActivity().finish();
        //
        //                    return true;
        //                }
        //
        //
        //                return false;
        //            }
        //        });

        //TODO shoudnt it be in onCreateView?

        val homeActivity = activity as HomeActivity?
        homeActivity?.fab?.visibility = View.VISIBLE
        homeActivity?.fab?.setImageResource(R.drawable.ic_add_24dp)
        homeActivity?.fab?.setOnClickListener { insertThread() }
    }

    private fun insertThread() {

        val uploadFragment = UploadFragment()

        val fragmentManager = (context as FragmentActivity).supportFragmentManager
        val transaction = fragmentManager.beginTransaction()
        transaction.replace(R.id.content_home, uploadFragment).addToBackStack(null)
        transaction.commit()

    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)

        if (isVisibleToUser && isResumed) {
            onResume()
        }
    }

    // TODO: Rename method, update argument and hook method into UI event
    fun onButtonPressed(uri: Uri) {
        if (listener != null) {
            listener?.onFragmentInteraction(uri)
        }
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is OnFragmentInteractionListener) {
            listener = context
            helper = Helper(context)

        } else {
            throw RuntimeException(context.toString() + " must implement OnFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    enum class LayoutManagerType {
        GRID_LAYOUT_MANAGER,
        LINEAR_LAYOUT_MANAGER
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     *
     *
     * See the Android Training lesson [Communicating with Other Fragments](http://developer.android.com/training/basics/fragments/communicating.html) for more information.
     */
    interface OnFragmentInteractionListener {
        // TODO: Update argument type and name
        fun onFragmentInteraction(uri: Uri)
    }

    companion object {
        val KEY_LAYOUT_MANAGER = "layoutManager"
        private val TAG = "HomeFragment"
    }
}// Required empty public constructor
